pkg/auth/server/selectprovider/selectprovider.go

```diff
@@ -0,0 +1,125 @@
+package selectprovider
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"html/template"
+	"net/http"
+
+	"github.com/openshift/origin/pkg/auth/oauth/handlers"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+const (
+	useRedirectHandlerParam = "useRedirectHandler"
+)
+
+type SelectProviderRenderer interface {
+	Render(redirectors []handlers.ProviderInfo, w http.ResponseWriter, req *http.Request)
+}
+
+type SelectProvider struct {
+	render            SelectProviderRenderer
+	forceInterstitial bool
+}
+
+var _ = handlers.AuthenticationSelectionHandler(&SelectProvider{})
+
+func NewSelectProvider(render SelectProviderRenderer, forceInterstitial bool) *SelectProvider {
+	return &SelectProvider{
+		render:            render,
+		forceInterstitial: forceInterstitial,
+	}
+}
+
+type ProviderData struct {
+	Providers []handlers.ProviderInfo
+}
+
+// NewSelectProviderRenderer creates a select provider renderer that takes in an optional custom template to
+// allow branding of the page. Uses the default if customSelectProviderTemplateFile is not set.
+func NewSelectProviderRenderer(customSelectProviderTemplateFile string) (*selectProviderTemplateRenderer, error) {
+	r := &selectProviderTemplateRenderer{}
+	if len(customSelectProviderTemplateFile) > 0 {
+		customTemplate, err := template.ParseFiles(customSelectProviderTemplateFile)
+		if err != nil {
+			return nil, err
+		}
+		r.selectProviderTemplate = customTemplate
+	} else {
+		r.selectProviderTemplate = defaultSelectProviderTemplate
+	}
+
+	return r, nil
+}
+
+func (s *SelectProvider) SelectAuthentication(providers []handlers.ProviderInfo, w http.ResponseWriter, req *http.Request) (handlers.ProviderInfo, bool, error) {
+	if len(providers) == 0 {
+		return handlers.ProviderInfo{}, false, nil
```

Comment:
yeah, I think returning `nil, false, nil` would be better here and in the `handled` case